### PR TITLE
Reworked header.

### PR DESCRIPTION
**Pull request by tommikas**
Function declarations always exist, but when tracing is disabled they are hidden by the dummy macros.

### CODE REVIEW

**Comment by tommikas** [comment]
@kjbracey-arm Is this about what you had in mind?

File: mbed-trace/mbed_trace.h [hunk 1]
```diff
@@ -108,9 +108,9 @@ extern "C" {
 #define tr_error(...)           mbed_tracef(TRACE_LEVEL_ERROR,   TRACE_GROUP, __VA_ARGS__)   //!< Print Error Message
 #define tr_err(...)             mbed_tracef(TRACE_LEVEL_ERROR,   TRACE_GROUP, __VA_ARGS__)   //!< Alternative error message
 #define tr_cmdline(...)         mbed_tracef(TRACE_LEVEL_CMD,     TRACE_GROUP, __VA_ARGS__)   //!< Special print for cmdline. See more from TRACE_LEVEL_CMD -level
-
-/** Possible to skip all traces in compile time */
-#if defined(YOTTA_CFG_MBED_TRACE) || (defined(YOTTA_CFG) && !defined(NDEBUG))
+//aliases for the most commonly used functions
+#define tracef(dlevel, grp, ...)    mbed_tracef(dlevel, grp, __VA_ARGS__)   //!< Alias for mbed_tracef()
+#define trace_array(buf, len)       mbed_trace_array(buf, len)              //!< Alias for mbed_tracef()
```
Comment (jupe) [inline mbed-trace/mbed_trace.h]: just thinking.. should there be `tr_array` macro for this also..

**Comment by jupe** [comment]
seems that you have to write some more tests because CI fails:
```
11:36:56 Code coverage enforcement failed for the following metrics:
11:36:56     Lines's stability is 65.41 and set mininum stability is 83.77.
11:36:56     Conditionals's stability is 50.56 and set mininum stability is 64.08
```

**Comment by tommikas** [comment]
Ah yes, that was actually something I did want comments on: Any other usage macros/aliases to add?

**Comment by tommikas** [comment]
Hm... I don't think the code coverage failure is because of this change. ARMmbed/mbed-trace#20 is failing for that too, no?

Still, regardless of what changes cause the failure, those tests should be added.

**Comment by tommikas** [comment]
Regarding the tr_array macro, I'm not sure it would add much. It isn't really possible to assign any standard values to the arguments of trace_array, like in the case of the other tr_ macros, so it would simply function as another alias for mbed_trace_array, saving 3 characters per line when compared to trace_array.

File: mbed-trace/mbed_trace.h (hunk 1 above)
Comment (kjbracey-arm) [inline mbed-trace/mbed_trace.h]: Would seem logical. (But as an addition for now, to avoid immediate respellings.)

File: mbed-trace/mbed_trace.h [hunk 2]
```diff
@@ -273,34 +276,41 @@ char* mbed_trace_ipv6_prefix(const uint8_t *prefix, uint8_t prefix_len);
...
-
-#endif // defined(__GNUC__) || defined(__CC_ARM)
+#ifdef __cplusplus
+}
+#endif
 
+#endif /* MBED_TRACE_H_ */
 
-#else // YOTTA_CFG_MBED_TRACE
+/* These macros are outside the inclusion guard so they will be re-evaluated for each individual compilation unit and
```
Comment (kjbracey-arm) [inline mbed-trace/mbed_trace.h]: Re-evaluated for each inclusion of the header - can include it more than once in a compilation unit.

File: mbed-trace/mbed_trace.h [hunk 3]
```diff
@@ -311,20 +321,14 @@ char* mbed_trace_ipv6_prefix(const uint8_t *prefix, uint8_t prefix_len);
...
-#define mbed_trace_last(...)                ((void) 0)
+#define mbed_trace_include_filters_get(...) ((char *) 0)
+#define mbed_trace_last(...)                ((char *) 0)
 #define mbed_tracef(...)                    ((void) 0)
-#define mbed_trace_ipv6(...)                ((void) 0)
-#define mbed_trace_array(...)               ((void) 0)
-#define mbed_trace_ipv6_prefix(...)         ((void) 0)
 
-#endif //YOTTA_CFG_MBED_TRACE
+#define mbed_trace_ipv6(...)                dont_use_trace_helpers_outside_trace_calls
```
Comment (kjbracey-arm) [inline mbed-trace/mbed_trace.h]: Maybe worth a comment explaining the restriction - they accumulate strings in a buffer that is only flushed by trace calls, and could overflow if used outside them.

File: mbed-trace/mbed_trace.h (hunk 1 above)
Comment (kjbracey-arm) [inline mbed-trace/mbed_trace.h]: What about trace_ipv6 and trace_ipv6_prefix?

File: mbed-trace/mbed_trace.h [hunk 4]
```diff
@@ -273,34 +276,41 @@ char* mbed_trace_ipv6_prefix(const uint8_t *prefix, uint8_t prefix_len);
...
+#endif
 
+#endif /* MBED_TRACE_H_ */
 
-#else // YOTTA_CFG_MBED_TRACE
+/* These macros are outside the inclusion guard so they will be re-evaluated for each individual compilation unit and
+ * if tracing is disabled, the dummies will hide the real functions. The real functions can still be reached by
+ * surrounding the name of the function with brackets, e.g. "(mbed_tracef)(dlevel, grp, "like so");"
+ * */
+#if defined(YOTTA_CFG_MBED_TRACE) || (defined(YOTTA_CFG) && !defined(NDEBUG))
```
Comment (kjbracey-arm) [inline mbed-trace/mbed_trace.h]: FEA_TRACE_SUPPORT going to go in via the separate commit, and then resolving the merge conflict?

**Comment by kjbracey-arm** [comment]
This is basically what I was thinking of, yes. Like it, just those few comments.

**Comment by tommikas** [comment]
>Re-evaluated for each inclusion of the header - can include it more than once in a compilation unit.

Yeah, realized this after pushing, heh.

>FEA_TRACE_SUPPORT going to go in via the separate commit, and then resolving the merge conflict?

That's what I was thinking.

**Comment by tommikas** [comment]
@kjbracey-arm. Incorporated your previous comments. Anything else come to mind?

**Comment by jupe** [comment]
lgtm

**Comment by tommikas** [comment]
test this please

**Comment by SeppoTakalo** [comment]
+1

**Comment by jupe** [comment]
lgtm
